✨ Allow template to change parser delimiters

ruanxin opened the pull request:
### Description
This change add `SetDelim()` and `GetDelim()` to `Template` interface, give the developer the ability to config an alternative delimiters (by default is `{{` and `}}`) for template parsing.

### Motivation
The current `controllerResourcesTemplate`, `controllerRuntimeTemplate` provided as a grafana dashboard with json format, and it contains some annotations syntax which use `{{ }}` and it's collide with default delimiters for template parsing. 

In order to parse correctly, at the moment, the template data have to write like this:
```
"legendFormat": "P90 {{"{{name}}"}} {{"{{instance}}"}} "
```
It is quite confusing to developers at first glance, and ever harder to do further encoding process like `json.Valid`, `json.Compact`.

By providing an alternative delimiters allow developer write template content in correct format without seeking workaround in order to deal with parse.

<!--

Hiya!  Welcome to Kubebuilder!  For a smooth PR process, please ensure
that you include the following information:

* a description of the change
* the motivation for the change
* what issue it fixes, if any, in GitHub syntax (e.g. Fixes #XYZ)

Both the description and motivation may reference other issues and PRs,
but should be mostly understandable without following the links (e.g. when
reading the git commit log).

Please don't @-mention people in PR or commit messages (do so in an
additional comment).

please add an icon to the title of this PR depending on the type:

- ⚠ (:warning:): breaking
- ✨ (:sparkles:): new non-breaking feature
- 🐛 (:bug:): bugfix
- 📖 (:book:): documentation
- 🌱 (:seedling:): infrastructure/other

See https://sigs.k8s.io/kubebuilder-release-tools for more information.

**PLEASE REMOVE THIS COMMENT BLOCK BEFORE SUBMITTING THE PR** (the bits
between the arrows)

-->

camilamacedo86 commented:
/ok-to-test


Thank you a lot for your contribution 🥇 
Could you please squash the commits

Kavinjsir commented:
Hi @ruanxin 👋🏼 
Glad to have your contribution! I think it is a good idea to resolve string literal`{{...}}` in go template.
This approach works for me locally. 👍🏼
Inject `[[`, `]]` seems no side effect for grafana template so far.

@camilamacedo86 @varshaprasad96 Would you help with a look? There are updates in `src/machinery`, mainly having additional `SetDelim` & `GetDelim` method, that works for the grafana plugin.

ruanxin commented:
/test pull-kubebuilder-test

ruanxin commented:
> Hi @ruanxin 👋🏼 Glad to have your contribution! I think it is a good idea to resolve string literal`{{...}}` in go template. This approach works for me locally. 👍🏼 Inject `[[`, `]]` seems no side effect for grafana template so far.
> 
> @camilamacedo86 @varshaprasad96 Would you help with a look? There are updates in `src/machinery`, mainly having additional `SetDelim` & `GetDelim` method, that works for the grafana plugin.

Hi @Kavinjsir, thank you! It's glad if I can contribute a bit on the kubebuilder improvement. Just to clarify, the `[[ ]]` pair I configured for those json format template is just an example to show how can we avoid that workaround solution, there is no any variable in those template need to be delimited at the moment, but if we need some variable later, it can be introduced like this `[[ .AnyVariable ]]`, and of cause we can change to other meaningful characters as delimiters.

everettraven reviewed (approved):
These changes look good and make sense to me.

I like that it enables us to not have to worry about potential delimiter overlaps for future templates.

Thanks for the contribution @ruanxin

camilamacedo86 commented on pkg/machinery/scaffold_test.go:
```diff
@@ -490,8 +490,19 @@ var _ Template = fakeTemplate{}
+func (f fakeTemplate) GetDelim() (string, string) {
+	return f.parseDelimLeft, f.parseDelimRight
 }
 
```
@ruanxin that seems great, 
Should we not add here some 1 or 2 use cases using the option to ensure that all work fine? 

WDYT?

ruanxin replied:
Hi @camilamacedo86, it's a default behavior of go template, I'm not sure if it's make too much sense to test it, but I added two test cases anyway.

camilamacedo86 commented on pkg/machinery/mixins.go:
```diff
@@ -48,14 +48,25 @@ type TemplateMixin struct {
 func (t *TemplateMixin) GetBody() string {
 	return t.TemplateBody
 }
 
```
Could we add comments to the new methods?
See that they are exported and the comments will be used to doc them.

camilamacedo86 commented on pkg/plugins/optional/grafana/v1alpha/scaffolds/internal/templates/resources.go:
```diff
@@ -34,7 +34,7 @@ func (f *ResourcesManifest) SetTemplateDefaults() error {
 		f.Path = filepath.Join("grafana", "controller-resources-metrics.json")
 	}
-
+	f.SetDelim("[[", "]]")
```
Could you please add on top a comment explaining why are you seeing the Delim and why the values adopt are ("[[", "]]"? So that, it makes easier for others to understand it in the future if they need it.

ruanxin replied:
done

camilamacedo86 commented on pkg/plugins/optional/grafana/v1alpha/scaffolds/internal/templates/runtime.go:
```diff
@@ -34,9 +34,8 @@ func (f *RuntimeManifest) SetTemplateDefaults() error {
 		f.Path = filepath.Join("grafana", "controller-runtime-metrics.json")
 	}
-
+	f.SetDelim("[[", "]]")
```
Same here,

ruanxin replied:
done

camilamacedo86 reviewed (commented):
@ruanxin that seems great 🥇 
I just added a few nit comments, could you please look them out?
Also, could you please squash the commits so we are able to get this PR merged?

ruanxin commented:
> @ruanxin that seems great 🥇 I just added a few nit comments, could you please look them out? Also, could you please squash the commits so we are able to get this PR merged?

Hi @camilamacedo86 , thanks for the comments! I resolved them and squashed my commits, please have check.

camilamacedo86 reviewed (approved):
That shows great for me too 🥇 

/approved

camilamacedo86 commented on pkg/machinery/interfaces.go:
```diff
@@ -44,6 +44,10 @@ type Template interface {
 	GetBody() string
 	// SetTemplateDefaults sets the default values for templates
```
@ruanxin, 

Could you please fix the CI issues so we are able to move forward? 

a) Lint: 

  Error: SA4005: ineffective assignment to field fakeTemplate.parseDelimLeft (staticcheck)
[27](https://github.com/kubernetes-sigs/kubebuilder/runs/8119443645?check_suite_focus=true#step:5:29)
  Error: SA4005: ineffective assignment to field fakeTemplate.parseDelimRight (staticcheck)
[2](https://github.com/kubernetes-sigs/kubebuilder/runs/8119443645?check_suite_focus=true#step:5:30)  
[29](https://github.com/kubernetes-sigs/kubebuilder/runs/8119443645?check_suite_focus=true#step:5:31)
  Error: issues found
[30](https://github.com/kubernetes-sigs/kubebuilder/runs/8119443645?check_suite_focus=true#step:5:32)
  Ran golangci-lint in 18854ms

Note: You can run make lint locally to check it out

b) The API one we just need to check / ensure that it does not break anything and we can move forward with it failing

c) the e2e test failures we need to ensure that are not related to this one

d) the coveralls we need to ensure that we add tests to pass on this one.

Could you please also rebase with the master to ensure that you have the latest changes?

ruanxin replied:
Hi @camilamacedo86 ,

I resolved this lint complain, it requires some refactoring. Just explain here, the `fakeTemplate` have to use pointer receiver, actually it's a golang best practice [If the method needs to mutate the receiver, the receiver must be a pointer.](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type), because of this, the struct has to be allocated as `&fakeTemplate` since the receiver is pointer `*fakeTemplate` now, all basic golang stuff, just want to clarify the change.

btw, I also rebased with the master.

camilamacedo86 commented:
/test pull-kubebuilder-e2e-k8s-1-19-16

camilamacedo86 commented:
/test pull-kubebuilder-e2e-k8s-1-24-1

camilamacedo86 commented:
/test pull-kubebuilder-e2e-k8s-1-21-2

camilamacedo86 commented:
/test-all

ruanxin commented:
/test pull-kubebuilder-e2e-k8s-1-20-7

ruanxin commented:
/test pull-kubebuilder-e2e-k8s-1-19-16

camilamacedo86 commented:
/test pull-kubebuilder-e2e-k8s-1-19-16
